pyairctrl/airctrl.py

```diff
@@ -11,10 +11,47 @@
 import sys
 import pprint
 import configparser
+import socket
+import xml.etree.ElementTree as ET
 
 G = int('A4D1CBD5C3FD34126765A442EFB99905F8104DD258AC507FD6406CFF14266D31266FEA1E5C41564B777E690F5504F213160217B4B01B886A5E91547F9E2749F4D7FBD7D3B9A92EE1909D0D2263F80A76A6A24C087A091F531DBF0A0169B6A28AD662A4D18E73AFA32D779D5918D08BC8858F4DCEF97C2A24855E6EEB22B3B2E5', 16)
 P = int('B10B8F96A080E01DDE92DE5EAE5D54EC52C99FBCFB06A3C69A6A9DCA52D23B616073E28675A23D189838EF1E2EE652C013ECB4AEA906112324975C3CD49B83BFACCBDD7D90C4BD7098488E9C219A73724EFFD6FAE5644738FAA31A4FF55BCCC0A151AF5F0DC8B4BD45BF37DF365C1A65E68CFDA76D4DA708DF1FB2BC2E4A4371', 16)
 
+def ssdp(timeout=1, repeats=3, debug=False):
+    addr = '239.255.255.250'
+    port = 1900
+    msg = '\r\n'.join([
+        'M-SEARCH * HTTP/1.1',
+        'HOST: {}:{}'.format(addr, port),
+        'ST: urn:philips-com:device:DiProduct:1',
+        'MX: 1', 'MAN: "ssdp:discover"','', '']).encode('ascii')
+    resp = []
+    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP) as s:
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+        s.setsockopt(socket.SOL_IP, socket.IP_MULTICAST_TTL, 20)
+        s.settimeout(timeout)
+        for i in range(repeats):
+            s.sendto(msg, (addr, port))
+            try:
+                while True:
+                    data, (ip, _) = s.recvfrom(1024)
+                    url = next((x for x in data.decode('ascii').splitlines() if x.startswith('LOCATION: ')), None)
```

Comment:
if you get the url you can exit the loop to avoid waiting for timeout

Reply:
ah, maybe this is how we handle the case when there are multiple purifiers in the network?

Reply:
I'm waiting for all responses in the timeout window to collect all active devices - the next line adds all discovered IP/URLs to response list

Reply:
Got it. Parsing the url is a bit convoluted but I guess I am fine for now. Consider using a dict that maps IP to location url rather than a list of lists.

Reply:
updated